python/lsst/pipe/tasks/matchBackgrounds.py

```diff
@@ -19,683 +19,627 @@
+        # TODO: leaving as-is for now, but currently this only fits the pre-
+        # binned images, meaning binSize is always 1.  But we need a solution
+        # to the 0 variance image issue, which might mean we need a variable
+        # binSize still.
+        nx = warp.getWidth() // binSize
+        ny = warp.getHeight() // binSize
 
+        bgCtrl = BackgroundControl(nx, ny, self.statsCtrl, self.statsFlag)
+        bgCtrl.setUndersampleStyle(self.config.undersampleStyle)
+        bkgd = makeBackground(warp, bgCtrl)
 
-class DataRefMatcher:
-    """Match data references for a specified dataset type.
+        return bkgd, bgCtrl
 
-    Note that this is not exact, but should suffice for this task
-    until there is better support for this kind of thing in the butler.
-
-    Parameters
-    ----------
-    butler : `lsst.daf.butler.Butler`
-        Butler to search for maches in.
-    datasetType : `str`
-        Dataset type to match.
-    """
-
-    def __init__(self, butler, datasetType):
-        self._datasetType = datasetType  # for diagnostics
-        self._keyNames = butler.getKeys(datasetType)
-
-    def _makeKey(self, ref):
-        """Return a tuple of values for the specified keyNames.
+    def _fluxScale(self, exposure):
+        """Scales image to nJy flux using photometric calibration.
 
         Parameters
         ----------
-        ref : `Unknown`
-            Data reference.
+        exposure: `lsst.afw.image._exposure.ExposureF`
+            Exposure to scale.
 
-        Raises
-        ------
-        KeyError
-            Raised if ref.dataId is missing a key in keyNames.
+        Returns
+        -------
+        fluxZp: `float`
+            Counts to nanojanskies conversion factor
         """
-        return tuple(ref.dataId[key] for key in self._keyNames)
-
-    def isMatch(self, ref0, ref1):
-        """Return True if ref0 == ref1.
-
-        Parameters
-        ----------
-        ref0 : `Unknown`
-            Data for ref 0.
-        ref1 : `Unknown`
-            Data for ref 1.
+        fluxZp = exposure.getPhotoCalib().instFluxToNanojansky(1)
+        exposure.image *= fluxZp
+        return fluxZp
 
-        Raises
-        ------
-        KeyError
-            Raised if either ID is missing a key in keyNames.
-        """
-        return self._makeKey(ref0) == self._makeKey(ref1)
+    @timeMethod
+    def matchBackgrounds(self, warps, skyMap, refVisitId):
+        """Match science exposures' background level to that of reference
+        exposure.
+
+        Process creates binned images of the full focal plane (in tract
+        coordinates) for all visit IDs, subtracts each from a similarly
+        binned FFP reference image, then generates TractBackground
+        objects.  It assumes (but does not require/check) that the mask planes
+        already have detections set.  If detections have not been set/masked,
+        sources will bias the difference image background estimation.
```

Comment:
Shouldn't the sources subtract out, and hence not bias the difference image background estimation?

If it looks like they _are_ biasing the backgrounds, then that sort of throws the whole premise into doubt (unless it's something simple, like needing to use PSF-matched warps instead of direct warps).